File: genshin-builder-app/src/lib/artifact-score.ts
/**
 * 聖遺物スコア計算
 *
 * キャラクターごとに「推奨スコア計算タイプ」を持ち、参照ステータスに応じて
 * 計算式を切り替える。新しい計算タイプを追加する場合は
 * SCORE_FORMULAS にエントリを足すだけでよい。
 *
 *   攻撃基準:   会心率×2 + 会心ダメージ + 攻撃力%
 *   HP基準:     会心率×2 + 会心ダメージ + HP%
 *   防御基準:   会心率×2 + 会心ダメージ + 防御力%
 *   熟知基準:   会心率×2 + 会心ダメージ + 元素熟知÷4
 */

import type {
  ArtifactPiece,
  ArtifactState,
  ArtifactSubstat,
} from "@/types/character";

export type ScoreType = "atk" | "hp" | "def" | "em";

export const SCORE_TYPE_LABEL: Record<ScoreType, string> = {
  atk: "攻撃力基準",
  hp: "HP基準",
  def: "防御力基準",
  em: "元素熟知基準",
};

/** タイプごとの「会心系以外」のスコア寄与計算 */
const SCORE_FORMULAS: Record<ScoreType, (sub: ArtifactSubstat) => number> = {
  atk: (sub) => (sub.stat === "攻撃力%" ? sub.value : 0),
  hp: (sub) => (sub.stat === "HP%" ? sub.value : 0),
  def: (sub) => (sub.stat === "防御力%" ? sub.value : 0),
  em: (sub) => (sub.stat === "元素熟知" ? sub.value / 4 : 0),
};

/**
 * キャラクター名 → 推奨スコア計算タイプ
 * 未登録のキャラクターは攻撃力基準（DEFAULT_SCORE_TYPE）を使う。
 * 新キャラクターはここに追記するだけでよい。
 */
const DEFAULT_SCORE_TYPE: ScoreType = "atk";

const SCORE_TYPE_BY_NAME: Record<string, ScoreType> = {
  // HP参照キャラクター
  胡桃: "hp",
  鍾離: "hp",
  珊瑚宮心海: "hp",
  夜蘭: "hp",
  ニィロウ: "hp",
  ディシア: "hp",
  白朮: "hp",
  フリーナ: "hp",
  ヌヴィレット: "hp",
  シグウィン: "hp",
  ムアラニ: "hp",
  // 防御力参照キャラクター
  ノエル: "def",
  荒瀧一斗: "def",
  ゴロー: "def",
  雲菫: "def",
  千織: "def",
  シロネン: "def",
  カチーナ: "def",
  // 元素熟知特化キャラクター
  楓原万葉: "em",
  スクロース: "em",
  ナヒーダ: "em",
  綺良々: "em",
  ヨォーヨ: "em",
  コレイ: "em",
  ティナリ: "em",
  久岐忍: "em",
  ラウマ: "em",
  アイノ: "em",
};

/** キャラクターの推奨スコア計算タイプを返す（マスターデータの scoreType を優先） */
export function getScoreType(
  character: { name: string; scoreType?: string | null },
): ScoreType {
  const fromDb = character.scoreType as ScoreType | undefined;
  if (fromDb && fromDb in SCORE_TYPE_LABEL) return fromDb;
  return SCORE_TYPE_BY_NAME[character.name] ?? DEFAULT_SCORE_TYPE;
}

/**
 * APIの specialProp からスコア計算タイプを推定する（同期時に使用）。
 * 名前ベースの上書き（胡桃=hp など）もここで適用する。
 */
export function inferScoreType(
  specialProp: string | null | undefined,
  name: string,
): ScoreType {
  if (SCORE_TYPE_BY_NAME[name]) return SCORE_TYPE_BY_NAME[name];

  switch (specialProp) {
    case "FIGHT_PROP_HP_PERCENT":
      return "hp";
    case "FIGHT_PROP_DEFENSE_PERCENT":
      return "def";
    case "FIGHT_PROP_ELEMENT_MASTERY":
      return "em";
    case "FIGHT_PROP_ATTACK_PERCENT":
      return "atk";
    default:
      return DEFAULT_SCORE_TYPE;
  }
}

/** 1部位分のスコアを計算する */
export function calcPieceScore(
  piece: ArtifactPiece,
  type: ScoreType = DEFAULT_SCORE_TYPE,
): number {
  let score = 0;
  for (const sub of piece.substats) {
    if (sub.stat === "会心率") score += sub.value * 2;
    else if (sub.stat === "会心ダメージ") score += sub.value;
    else score += SCORE_FORMULAS[type](sub);
  }
  return Math.round(score * 10) / 10;
}

/** 5部位合計のスコアを計算する */
export function calcTotalScore(
  artifacts: ArtifactState,
  type: ScoreType = DEFAULT_SCORE_TYPE,
): number {
  const total = Object.values(artifacts).reduce(
    (sum, piece) => sum + calcPieceScore(piece, type),
    0,
  );
  return Math.round(total * 10) / 10;
}

File: genshin-builder-mobile/lib/domain/artifact_score.dart
import 'artifact_score_weights.dart';
import 'models/master_models.dart';
import 'models/artifact_state.dart';

enum ArtifactScoreType { atk, hp, def, recharge, em }

const _defaultScoreType = ArtifactScoreType.atk;

bool isGenericMasterScoreType(String? raw) {
  final normalized = raw?.trim().toLowerCase();
  return normalized == null || normalized.isEmpty || normalized == 'atk';
}

ArtifactScoreType resolveArtifactScoreType(
  MasterCharacter character, {
  Map<String, ArtifactScoreType>? nameOverrides,
}) {
  final fromDb = artifactScoreTypeFromString(character.scoreType);
  if (fromDb != null && !isGenericMasterScoreType(character.scoreType)) {
    return fromDb;
  }

  final fromName = nameOverrides?[character.name];
  if (fromName != null) return fromName;

  return fromDb ?? _defaultScoreType;
}

/// 同期時に Amber の specialProp と名前から scoreType を推定する。
/// Web版 `inferScoreType` と同等（recharge 対応を追加）。
/// [nameOverrides] は `artifact_score_type_overrides.json` を読み込んだマップ。
ArtifactScoreType inferScoreType(
  String? specialProp,
  String name, {
  Map<String, ArtifactScoreType>? nameOverrides,
}) {
  final fromName = nameOverrides?[name];
  if (fromName != null) return fromName;

  return switch (specialProp) {
    'FIGHT_PROP_HP_PERCENT' => ArtifactScoreType.hp,
    'FIGHT_PROP_DEFENSE_PERCENT' => ArtifactScoreType.def,
    'FIGHT_PROP_ELEMENT_MASTERY' => ArtifactScoreType.em,
    'FIGHT_PROP_CHARGE_EFFICIENCY' => ArtifactScoreType.recharge,
    'FIGHT_PROP_ATTACK_PERCENT' => ArtifactScoreType.atk,
    _ => _defaultScoreType,
  };
}

ArtifactScoreType? artifactScoreTypeFromString(String? raw) => switch (raw) {
  'atk' => ArtifactScoreType.atk,
  'hp' => ArtifactScoreType.hp,
  'def' => ArtifactScoreType.def,
  'recharge' || 'er' => ArtifactScoreType.recharge,
  'em' => ArtifactScoreType.em,
  _ => null,
};

String artifactScoreTypeToStorage(ArtifactScoreType type) => switch (type) {
  ArtifactScoreType.atk => 'atk',
  ArtifactScoreType.hp => 'hp',
  ArtifactScoreType.def => 'def',
  ArtifactScoreType.recharge => 'recharge',
  ArtifactScoreType.em => 'em',
};

const _userArtifactScoreTypePrefix = 'user:';

/// ユーザーが手動選択したスコア基準のみ永続化する。
String artifactScoreTypeToUserStorage(ArtifactScoreType type) =>
    '$_userArtifactScoreTypePrefix${artifactScoreTypeToStorage(type)}';

ArtifactScoreType? userArtifactScoreTypeFromStorage(String? raw) {
  if (raw == null || !raw.startsWith(_userArtifactScoreTypePrefix)) {
    return null;
  }
  return artifactScoreTypeFromString(
    raw.substring(_userArtifactScoreTypePrefix.length),
  );
}

ArtifactScoreType? inferArtifactScoreTypeFromWeights(
  ArtifactStatWeights weights,
) {
  for (final type in ArtifactScoreType.values) {
    if (_sameWeights(weights, scoreWeightsForType(type))) {
      return type;
    }
  }
  return null;
}

double calcArtifactPieceScore(ArtifactPiece piece, ArtifactScoreType type) {
  return calcArtifactPieceScoreWithWeights(piece, scoreWeightsForType(type));
}

ArtifactStatWeights scoreWeightsForType(ArtifactScoreType type) =>
    switch (type) {
      ArtifactScoreType.atk => const ArtifactStatWeights(
        critRate: 2,
        critDamage: 1,
        atkPercent: 1,
        hpPercent: 0,
        defPercent: 0,
        elementalMastery: 0,
        energyRecharge: 0,
      ),
      ArtifactScoreType.hp => const ArtifactStatWeights(
        critRate: 2,
        critDamage: 1,
        atkPercent: 0,
        hpPercent: 1,
        defPercent: 0,
        elementalMastery: 0,
        energyRecharge: 0,
      ),
      ArtifactScoreType.def => const ArtifactStatWeights(
        critRate: 2,
        critDamage: 1,
        atkPercent: 0,
        hpPercent: 0,
        defPercent: 1,
        elementalMastery: 0,
        energyRecharge: 0,
      ),
      ArtifactScoreType.recharge => const ArtifactStatWeights(
        critRate: 2,
        critDamage: 1,
        atkPercent: 0,
        hpPercent: 0,
        defPercent: 0,
        elementalMastery: 0,
        energyRecharge: 1,
      ),
      ArtifactScoreType.em => const ArtifactStatWeights(
        critRate: 2,
        critDamage: 1,
        atkPercent: 0,
        hpPercent: 0,
        defPercent: 0,
        elementalMastery: 0.25,
        energyRecharge: 0,
      ),
    };

double calcArtifactPieceScoreWithWeights(
  ArtifactPiece piece,
  ArtifactStatWeights weights,
) {
  var score = 0.0;
  for (final sub in piece.substats) {
    switch (sub.stat) {
      case '会心率':
        score += sub.value * weights.critRate;
        break;
      case '会心ダメージ':
        score += sub.value * weights.critDamage;
        break;
      case '攻撃力%':
        score += sub.value * weights.atkPercent;
        break;
      case 'HP%':
        score += sub.value * weights.hpPercent;
        break;
      case '防御力%':
        score += sub.value * weights.defPercent;
        break;
      case '元素チャージ効率':
        score += sub.value * weights.energyRecharge;
        break;
      case '元素熟知':
        score += sub.value * weights.elementalMastery;
        break;
      default:
        break;
    }
  }
  return _round1(score);
}

double calcArtifactTotalScore(ArtifactState artifacts, ArtifactScoreType type) {
  return calcArtifactTotalScoreWithWeights(
    artifacts,
    scoreWeightsForType(type),
  );
}

double calcArtifactTotalScoreWithWeights(
  ArtifactState artifacts,
  ArtifactStatWeights weights,
) {
  var total = 0.0;
  for (final slot in ArtifactSlotKey.values) {
    total += calcArtifactPieceScoreWithWeights(
      artifacts[slot] ?? createEmptyArtifactPiece(),
      weights,
    );
  }
  return _round1(total);
}

double _round1(double value) => (value * 10).roundToDouble() / 10;

bool _sameWeights(ArtifactStatWeights a, ArtifactStatWeights b) {
  const epsilon = 0.000001;
  bool close(double x, double y) => (x - y).abs() < epsilon;
  return close(a.critRate, b.critRate) &&
      close(a.critDamage, b.critDamage) &&
      close(a.atkPercent, b.atkPercent) &&
      close(a.hpPercent, b.hpPercent) &&
      close(a.defPercent, b.defPercent) &&
      close(a.elementalMastery, b.elementalMastery) &&
      close(a.energyRecharge, b.energyRecharge);
}
